gobblin-runtime/src/main/java/gobblin/runtime/job_catalog/CachingJobCatalog.java

```diff
@@ -0,0 +1,91 @@
+/*
+ * Copyright (C) 2014-2016 LinkedIn Corp. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use
+ * this file except in compliance with the License. You may obtain a copy of the
+ * License at  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed
+ * under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
+ * CONDITIONS OF ANY KIND, either express or implied.
+ */
+package gobblin.runtime.job_catalog;
+
+import java.net.URI;
+import java.util.Collection;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.common.base.Optional;
+
+import gobblin.runtime.api.JobCatalog;
+import gobblin.runtime.api.JobCatalogListener;
+import gobblin.runtime.api.JobSpec;
+
+/**
+ * A JobCatalog decorator that caches all JobSpecs in memory.
+ *
+ */
+public class CachingJobCatalog implements JobCatalog {
+  protected final JobCatalog _fallback;
+  protected final InMemoryJobCatalog _cache;
+  protected final Logger _log;
+
+  public CachingJobCatalog(JobCatalog fallback, Optional<Logger> log) {
+    _log = log.isPresent() ? log.get() : LoggerFactory.getLogger(getClass());
+    _fallback = fallback;
+    _cache = new InMemoryJobCatalog(log);
+    _fallback.addListener(new FallbackCatalogListener());
+  }
+
+  /** {@inheritDoc} */
+  @Override
+  public Collection<JobSpec> getJobs() {
+    return _cache.getJobs();
```

Comment:
Should this block until the fallback finishes notifying the cache layer of jobs? Otherwise the return value is not actually representative of `fallback.getJobs()`.